include/ensmallen_bits/ada_sqrt/ada_sqrt_update.hpp

```diff
@@ -0,0 +1,123 @@
+ *
+ * @code
+ * @misc{hu2019secondorder,
+ *   title  = {Second-order Information in First-order Optimization Methods},
+ *   author = {Yuzheng Hu and Licong Lin and Shange Tang},
+ *   year   = {2019},
+ *   eprint = {1912.09926},
+ * }
+ * @endcode
+ *
+ */
+class AdaSqrtUpdate
+{
+ public:
+  /**
+   * Construct the AdaSqrt update policy with given epsilon parameter.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   */
+  AdaSqrtUpdate(const double epsilon = 1e-8) : epsilon(epsilon), iteration(0)
+  {
+    // Nothing to do.
+  }
+
+  //! Get the value used to initialise the squared gradient parameter.
+  double Epsilon() const { return epsilon; }
+  //! Modify the value used to initialise the squared gradient parameter.
+  double& Epsilon() { return epsilon; }
+
+  //! Get the current iteration number.
+  size_t Iteration() const { return iteration; }
+  //! Modify the current iteration number.
+  size_t& Iteration() { return iteration; }
+
+  /**
+   * The UpdatePolicyType policy classes must contain an internal 'Policy'
+   * template class with two template arguments: MatType and GradType.  This is
+   * instantiated at the start of the optimization, and holds parameters
+   * specific to an individual optimization.
+   */
+  template<typename MatType, typename GradType>
+  class Policy
+  {
+   public:
+    /**
+     * This constructor is called by the SGD optimizer before the start of the
+     * iteration update process. In AdaSqrt update policy, squared gradient
+     * matrix is initialized to the zeros matrix with the same size as gradient
+     * matrix (see ens::SGD<>).
+     *
+     * @param parent Instantiated parent class.
+     * @param rows Number of rows in the gradient matrix.
+     * @param cols Number of columns in the gradient matrix.
+     */
+    Policy(AdaSqrtUpdate& parent, const size_t rows, const size_t cols) :
+        parent(parent),
+        squaredGradient(rows, cols)
+    {
+      // Initialize an empty matrix for sum of squares of parameter gradient.
+      squaredGradient.zeros();
+    }
+
+    /**
+     * Update step for SGD. The AdaSqrt update adapts the learning rate by
+     * performing larger updates for more sparse parameters and smaller updates
+     * for less sparse parameters.
+     *
+     * @param iterate Parameters that minimize the function.
+     * @param stepSize Step size to be used for the given iteration.
+     * @param gradient The gradient matrix.
+     */
+    void Update(MatType& iterate,
+                const double stepSize,
+                const GradType& gradient)
+    {
+      ++parent.iteration;
+
+      squaredGradient += (gradient % gradient);
+
+      iterate -= stepSize * std::sqrt(parent.iteration) * gradient /
+          (squaredGradient + parent.epsilon);
+    }
+
+   private:
+    // Instantiated parent class.
+    AdaSqrtUpdate& parent;
+    // The squared gradient matrix.
+    GradType squaredGradient;
```

Comment:
Up to you but it may be clearer to name this `sumSquaredGradients`.  I got a little confused and had to double-check the paper to understand.